### PR TITLE
Cv 677 suppress emails to idams absentees

### PR DESCRIPTION
**Pull request by chrisfoster76**

### CODE REVIEW

File: src/SFA.DAS.PAS.Account.Api/Orchestrator/EmailOrchestrator.cs [hunk 1]
```diff
@@ -15,65 +17,111 @@ public class EmailOrchestrator
...
+                _logger.Error(ex, "An error occurred retrieving users from Provider IDAMS");
+                idamsUsersTask = Task.FromResult(new List<string>());
+                idamsSuperUsersTask = Task.FromResult(new List<string>());
             }
 
+            var idamsUsers = await idamsUsersTask;
+            var idamsSuperUsers = await idamsSuperUsersTask;
+            var allIdamsUsers = idamsUsers.Concat(idamsSuperUsers).Distinct().ToList();
+            _logger.Info($"{allIdamsUsers.Count} total users retrieved from IDAMS for Provider {providerId} ({idamsUsers.Count} DAS Users; {idamsSuperUsers.Count} Super Users)");
+
```
Comment (Pauljgraham) [inline src/SFA.DAS.PAS.Account.Api/Orchestrator/EmailOrchestrator.cs]: The code just above here I found hard to determine what it was doing. Could we extract it into a function such as 

``` 
private async Task<(bool Error, List<string> Users)> GetIdamsUsers(long providerId)
        {
            Task<List<string>> idamsUsersTask;
            Task<List<string>> idamsSuperUsersTask;

            var idamsError = false;
            try
            {
                _logger.Info($"Retrieving DAS Users and Super Users from Provider IDAMS for Provider {providerId}");
                idamsUsersTask = _idamsEmailServiceWrapper.GetEmailsAsync(providerId);
                idamsSuperUsersTask = _idamsEmailServiceWrapper.GetSuperUserEmailsAsync(providerId);

                await Task.WhenAll(idamsUsersTask, idamsSuperUsersTask);
            }
            catch (Exception ex)
            {
                idamsError = true;
                _logger.Error(ex, "An error occurred retrieving users from Provider IDAMS");
                idamsUsersTask = Task.FromResult(new List<string>());
                idamsSuperUsersTask = Task.FromResult(new List<string>());
            }

            var idamsUsers = await idamsUsersTask;
            var idamsSuperUsers = await idamsSuperUsersTask;
            var allIdamsUsers = idamsUsers.Concat(idamsSuperUsers).Distinct().ToList();
            _logger.Info($"{allIdamsUsers.Count} total users retrieved from IDAMS for Provider {providerId} ({idamsUsers.Count} DAS Users; {idamsSuperUsers.Count} Super Users)");
            return (idamsError, allIdamsUsers)l
        } 
```

It gives an explicit intent to the action being performed.

File: src/SFA.DAS.PAS.Account.Api/Orchestrator/EmailOrchestrator.cs [hunk 2]
```diff
@@ -15,65 +17,111 @@ public class EmailOrchestrator
...
+            var accountUsers = (await _accountOrchestrator.GetAccountUsers(providerId)).ToList();
+
+            if (!idamsError) {  await RemoveAccountUsersNotInIdams(accountUsers, allIdamsUsers); }
```
Comment (Pauljgraham) [inline src/SFA.DAS.PAS.Account.Api/Orchestrator/EmailOrchestrator.cs]: I can see we are marking users who have been removed from Idams here. But what happens is the user is removed from Idams and then re-added? Do we undelete those users. Or will the normal options where opt out/in get re-applied?

File: src/SFA.DAS.PAS.Account.Api/Orchestrator/EmailOrchestrator.cs [hunk 3]
```diff
@@ -15,65 +17,111 @@ public class EmailOrchestrator
...
+                }
             }
 
-            recipients = await _idamsEmailServiceWrapper.GetSuperUserEmailsAsync(ukprn);
-            if (recipients.Any())
+            foreach (var user in removedUsers)
             {
-                return recipients;
+                _logger.Info($"Account User {user.UserRef} not found in IDAMS and will be marked as deleted");
+                await _mediator.Send(new DeleteRegisteredUserCommand { UserRef = user.UserRef });
```
Comment (Pauljgraham) [inline src/SFA.DAS.PAS.Account.Api/Orchestrator/EmailOrchestrator.cs]: Considering we have a list of users to remove would it be better to remove them as a single call

Comment (chrisfoster76) [reply src/SFA.DAS.PAS.Account.Api/Orchestrator/EmailOrchestrator.cs]: I think removal from IDAMS will be relatively uncommon (maybe not first time this code runs, but from then on).

File: src/SFA.DAS.PAS.Account.Api/Orchestrator/EmailOrchestrator.cs (hunk 2 above)
Comment (chrisfoster76) [reply src/SFA.DAS.PAS.Account.Api/Orchestrator/EmailOrchestrator.cs]: They get undeleted.

File: src/SFA.DAS.PAS.Account.Api/Orchestrator/EmailOrchestrator.cs (hunk 1 above)
Comment (chrisfoster76) [reply src/SFA.DAS.PAS.Account.Api/Orchestrator/EmailOrchestrator.cs]: Done!

**Pull request merged**